readme.md
Fork of [a fork](http://www.raspberrypi.org/phpBB3/viewtopic.php?f=78&t=23550) of [Osmose](http://bcz.asterope.fr/) for my [beagleboy](https://github.com/bear24rw/beagleboy) project.

--
```
           _______                  _______
          |       |.-----.--------.|       |.-----.-----.
          |   -   ||__ --|        ||   -   ||__ --|  -__|
          |_______||_____|__|__|__||_______||_____|_____|
           _______                  __         __
          |    ___|.--------.--.--.|  |.---.-.|  |_.-----.----.
          |    ___||        |  |  ||  ||  _  ||   _|  _  |   _|
          |_______||__|__|__|_____||__||___._||____|_____|__|
       Version 0.8.0 - 8th public release, written by Bruno Vedder.
```

Osmose means:
-------------

Object      Oriented
       S.m.s.         Emulator.

In brief it's an Sega Master System / Game Gear emulator encapsulated into C++
classes :-)


24 December 2006:  Osmose-0-8-1 9th public release.
--------------------------------------------------

Osmose is another Sega Master System / Gamegear emulator. The emulator is
released as open source project, under GPL license. For more information just
read the file 'license.txt'.

The emulator relies on SDL library and Zlib and Gilles Vollant unzip package.
Scale2x video Filter is based on Andrea Mazzoleni's algorithm.

What's new:
-----------

-Add Save state. User can save in 5 different slot emulation state
 and reload them later. Use keypad + - to change slot. F11 Load a state,
 F12 save a state in current slot.

Features:
--------

-SMS: Good compatibility. At this stage, the emulator can run* 96% of commercial
 games and public demos, except games that relies on codemaster mapper, which work
 but does not have proper video mode emulated.
-Game Gear: Good compatibility. At this stage, the emulator can run 98.0%* of
 game gear ROMS.
-SN76489 Sound is supported, but needs more work on noise register.
-support. for .zip .sms  and .gg format.
-Video filters: monochrom, scanline, scale2x, bilinear, nearest neighbour,
 and a new interesting scale2xscanline (fullscreen option always available).
-Pad(keyboard or joystick mapped), Sport Paddle (mouse mapped) emulation.
-PAL/NTSC Timing.
-Japanese/Export console.
-In game Screenshots, GFX rip, sound shot.
-Accurate auto-frameskip.
-Configurable keyboard configuration.
-Joystick support.

*Due to the huge number of game gear/sms (around 1300) roms, games have not been
deeply tested.

About scale2xscanline video filter: I was pleased by the scale2x renderer except
for it's brightness, to high for me. But i liked the rounded shapes instead of big
square pixels. That's why i wrote an alternate video filter which apply scale2x AND
scanline video filter. I find the result really interesting.

I finally add 2x Nearest neighbour video filter because some hardcore emulator users
don't accept (in term of emulation) screen modifier saying: real SMS does not have
bilinear video filtering. That's a point of view that i understand. So i add this
basic and pixel harmless video filter :)

Usage:
------

./osmose rom_name.zip -tv
./osmose test.sms -nosound -fs -tv
./osmose crazy_game.gg -inifile my_conf.ini

Notes:
ROM name MUST be the first parameter.
If you are not easy with command line, maybe should you use 'wxosmose', which
is a graphical frontend for Osmose. It's written by Bertram, thanks to him :-)


Options:
--------

Osmose 0-8-0 now implements embedded database, with game that needs specific
options to turn them on autmatically. User don't need to use -cm, -km -pal
or -jap options. But i don't own all roms, that why these option are still
usable from command line.

    -paddle           emulates one axis paddle (mapped on mouse).
    -joy              use joystick as input device, instead of keyboard.
    -acceleration x.x paddle acceleration (0.1 to 5 default 0.5)
    -fs               run in fullscreen   (default: windowed).
    -nosound          do not play sounds. (default: sound on).
    -dp               use dark palette for screen emulation (default: off).
    -inifile          xxx use xxx as configuration file.
    -fps              display fps in title bar.
    -monochrom        emulates B&W tv (default: off).
    -scale2x          scale2x video filter implementation (default: off).
    -bilinear         bilinear video filter (default: off).
    -tv               emulates TV scanline video filter (default: off).
    -scale2xscanline  Scale2x + 75% scanline video filter (default: off).
    -nn2x             nearest neighbour video filter (default: off).
    -cm               use codemaster games mem. mapper (default: off).
    -km               use korean games mem. mapper (default: off).
    -irqhack          Enable irq hack (specific rom option. default off).
    -pal              emulates PAL/SECAM video timing (default: NTSC).
    -jap              run as japanese sms (default: export).
    -exp              run as exported sms (default).

Important notes:

       -You can combine -tv/-monochrom/-scale2x with -fs (fullscreen).
       -If your config file name contains some spaces chars, use -inifile
        like this: -inifile "Name with space.txt" (with double quotes).

Default Keys:
-------------

SOUNDSHOT :  F1 (start/stop audio recording (22050kz 16bits/mono into wav file)
             Be carefull, wave file grows about 2.58 Mo/minute.
PAUSE     :  p
SCREENSHOT:  F2
TILESHOT  :  F3  (this will rip gfx used at this moment).
QUIT      :  ESCAPE
[DEBUGGER : d, if Osmose compiled with builtin debugger]

PAD1 UP   :  UP
PAD1 DOWN :  DOWN
PAD1 LEFT :  LEFT
PAD1 RIGHT:  RIGHT
PAD1 A    :  LEFT ALT
PAD1 B    :  LEFT CONTROL

PAD2 UP   :  NUMPAD 5
PAD2 DOWN :  NUMPAD 2
PAD2 LEFT :  NUMPAD 1
PAD2 RIGHT:  NUMPAD 3
PAD2 A    :  n
PAD2 B    :  b

ENTER     :  Game Gear start button.

KEYPAD   +:  Increment save state slot.
KEYPAD   -:  Decrement save state slot.
F11       :  Load a state from current slot.
F12       :  Save a state in current slot.

Keyboard configuration:
-----------------------

If you dislike Osmose's keyboard configuration, you can change it,
by providing a file with new configuration. When run, osmose will create
a osmose.ini file with default configuration. Under Linux/Unix this file
is created in ./osmose of user directory. You can specify other configuration
file with '-inifile my_config.txt' option

Sound shots, tiles rip and screenshots, Battery backed ram:
---------------------------------------

Osmose save these files in the corresponding folders:
./snd
./tiles
./screen
./bbr
./saves

If the emulator doesn't find osmose.ini, it will generate a new one, with
these four folders. Without these folders no save are possible. If Osmose
can't create them do it manually !

To force osmose to create these directory again, simply delete osmose.ini
and launch a ROM.

Building Osmose from sources:
-----------------------------

Assuming your system is correctly configured, with SDL library,
just type: make from the source directory.

Thanks to:
----------
-Enik, for bug reports, beta testing, and sound comments.

-Bertram, for the Osmose's Frontend, based on wxWidgets, and bug reports.

-Alessandro Scotti, for CPU core, and nice information about it.
 Web site: http://ascotti.org/programming/tickle/tickle.htm

-Charles MacDonald, for his excellent technical documentation on the SMS
 hardware.
 Web site: http://cgfm2.emuviews.com/

-SMS Power dev forum, for help about a lot of SMS parts, and nice resources.
 Web site: http://www.smspower.org/

-Zlib home page for their excellent library.
 Web site: http://www.gzip.org/zlib

-Gilles Vollant for it's unzip package based on zlib.

-emu-france.com for free and publicity free web hosting.
 Web site: http://emu-france.com

-Andrea Mazzoleni, for providing scale2x video filter algorithm.
 Web site: http://scale2x.sourceforge.net/

-Sound beta tester from 6t-web.
 Web site: http://6t-web.com

Feedback for bugs, problems or features requests are obiously welcome.

--------------------------------------------------------------------
|  Contact: osmose_emulator@yahoo.fr                               |
|  Site   : http://bcz.emu-france.com                              |
--------------------------------------------------------------------


changes.txt
Osmose-0-8-1
------------
-Add Save state. User can save in 5 different slot emulation state
 and reload them later. Use keypad + - to change slot. F11 Load a state,
 F12 save a state in current slot.

Osmose-0-8-0
------------
-Implement support for zoomed sprites.
	Corrected games:
	Nba_jam.gg
	doraemon_waku_waku_pocket_paradise-gg-jp.gg
	VR-trooper
	X-Men - Mojo World
	Virtua Fighter mini
	Earthworm jim (SMS and GG) uses 16*32 sprite :/
	And a few other one.
-Add option -irqhack.
-Add support for Korean Memory mapper. Mappers are now method-pointer driven instead
 of selected in wr8() with if/else condition.

Osmose-0-7-9
------------
-Add row 24-31 scrolling lock.
-Add embedded database with CRC of games that need specific options.
 User don't need to look for good option anymore.
-Add INTERRUPT HACK in vdp updateIRQAssertion. Some game need to run a few
 cycles before VDP generates interrupt. This Hack break a lot of games that
 worked before, so now it's an option.

Osmose-0-7-8
------------
-Modify the Debugger. It's now based on event.
-Add a lot of king of debug event (On @write, EI, DI, VDP IRQ asserted ect...)
-Correct page 0 banking. ROM at 0x000 - 0x400 is no more swapped if page 0 is
 changed.
 Spacegun now works.
-Add cpu->step() just before generating IRQ. This solve a lot of came lockup.
 Thanks to Charles Macdonalds for the trick.
	Chicago Syndicate
	Zool (SMS/GG)
	Monster Truck War
	Sega Chess

Osmose-0-7-7:
-------------
-Move IRQ generation into VDP update method instead of osmose core.

 IRQ generation is more accurate now.
 New game working:
	Castelo Ra Tin Bum (Brazil) [b1].sms*
	Castelo Ra Tin Bum (Brazil) [!].sms*


Osmose 0-7-6:
-------------
-Add josytick support. (-joy option)
-Simplification of input handling in OsmoseCore.cpp. It's faster.

Osmose 0-7-5:
-------------
-Remove IRQ control from OsmoseCore. It's now entirely handled by vdp.
-make port 0x2 return what was written.
-New game working:

  Primal Rage GG
  Stargate GG
  Pac Attack GG
  Mortal Combat II GG
  Griffin GG
  Crayon Chin Chan GG

Osmose 0-7-4:
-------------
-Modify code to avoid all GCC 4.1.1 warnings.
-Major speedup on Bilinear video filter.
-Remove -time_limite option.
-Add IRQ as debugger trigger.
-Totaly Rewrote VDP IRQ system. This correct one of the oldest bug in Osmose.
 A lot of games are no more flickering new working game:
  Joe Montana's Football GG
  GP Rider

New games working:
	Spider-Man - Return of the Sinister Six (JUE).gg
	Spider-Man - Return of the Sinister Six (JUE).sms
    Last Action Hero (UE)[!].gg
    Star Wars (collision detection).

Osmose 0-7-3:
-------------
-Adapt JR cpu Opcode (0x18) to GCC 4.1.1:
    PC += (s8)env.rd8(PC); PC++; 	work.
    PC += (s8)env.rd8(PC++);		does not work anymore.


Osmose 0-7-2:
-------------
-New SN76489.
-New Video filter: scale2xscanline (-scale2xscanline option)
-New Video filter: Nearest neighbour (-nn2x)

Osmose 0-7-1: (Major Change with GG suport).
-------------
-Add GAMEGEAR support.


Osmose 0-7-0: (Major Change New CPU).
-------------
-Switch to a new Z80 cpu core.


Osmose 0-6-30 Final Version:
----------------------------
-Correct a nasty bug in Paddle emulation. Now woody pop and megumi's revenge
 works smoothly, but the options -paddle -jap are REQUIRED.
-(Windows) Screenshots, tiles rip, sound rip and .bbr file are now stored into their own subdirectory.
-Few modification on windows frontend.
-Few code cleanup

Osmose 0-6-29:
--------------
-(Linux) Screenshots, tiles rip, sound rip and .bbr file arte now stored into their own subdirectory.
 This is much cleaner than puting all things together in home directory.

Osmose 0.6.28:
--------------
-Added support for paddle contrle device (-paddle option).
 This device is mapped on MOUSE, and MOUSE BUTTONS, keyboard is unused
 with this option, except general emulation keys, like escape, F1, F2 ect...
-Added -acceleration option. This option takes a parameter ex:
 example:   -acceleration 0.78


Osmose 0.6.27:
--------------
-Added support for Battery Backed memory. It's save when leaving emulator,
 and restore on startup, if file exists. The file is rom_name.bbr and is
 32768 bytes.
-Added support for vdp_status sprite collision bit5.
 Impossible mission seems fully playable (-pal option required).

Osmose 0.6.26:
--------------
-Implement bilinear filtering (-bilinear option).

Osmose 0.6.25:
--------------
-Modify OsmoseCore to handle "over cycles". When cpu core executes more
 than 228 Cycle during a scanline, over cycles are substract to CPU_CYCLE_PER_LINE.
 This should improve emulation timing.
 *Impossible mission is now playable (with -pal option), bug still buggy.

Osmose 0.6.24:
--------------
-Implement text writer, to display messages on emulator screen during
 emulation, instead of console message. This allow to read them, in
 fullscreen mode, and on windows platform.
-Add Key pad + - for volume control.
-Correct a bug in VDP that cause line traced when display disable to
 be not cleared. (This correct smurf 2 2nd level problem).

Osmose 0.6.23:
--------------
-Modify memory mapper to handle zip file containing more than just
 one .sms archive. Now files in archive are browsed and skipped if
 not with .sms extension. The first .sms or .SMS file found is used
 as our ROM.
-Modify fps display. If fps >= 60.0 display is "fps >= 60.0 (xxx.xx)
 if fps < 60.0  fps is displayed like this: "fps = xx.x"

Osmose 0.6.22:
--------------
-Modify OsmoseCore to handle NTSC or PAL/SECAM timings. (-pal option)
   *unsigned char *v_cnt points vcount_ntsc_192 or vcount_pal_192 and
    is used instead of only vcount_ntsc_192[]
   *VDP constructor initialises it based on option.ntsc value.
Games that works now:
  Back to the Future III
  A lot of other games had flickering problem fixed.
-Modify screenshot, soundshot and tile rip naming convention.
-Minor OsmoseCore code reorganisation.
-Set CPU_CYCLES_PER_LINE to 228 instead of 227.

Osmose 0.6.21:
--------------

main.cpp
/*
    Written on 2004 by Bruno Vedder

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/*****************************************************************************
*
* File: main.cpp
*
* Project: Osmose emulator.
*
* Author: Vedder Bruno
* Date: 28/10/2004, 13h00
*
* URL: http://bcz.emu-france.com/
*****************************************************************************/

#include <iostream>
#include <iomanip>
#include "OsmoseCore.h"
#include "Options.h"
#include "Version.h"
#include "VideoFilterList.h"

using namespace std;


// Functions prototypes.
void usage();
int parseCommandLine(int a, char *rv[]);
void splashText();

Options              opt;        // SMS Machine related options.
EmulatorOptions      emu_opt;    // Emulator related options.

/*--------------------------------------------------------------------*/
/* This is the main method of osmose. It check args number, correct   */
/* Emulator options, and start emulation is Ok.                       */
/*--------------------------------------------------------------------*/
int main(int argc,char *argv[])
{
    int ret;

    // Reset default options.
    emu_opt.reset();
    opt.reset();
    
    // Display Osmose version, author.
    splashText();

    // Check arguments number. If not enough display usage.
    if (argc < 2)
    {
        usage();
        exit(0);
    }

    // emu_opt will be corrected depending on cmd line.
    ret = parseCommandLine(argc, argv);

    if (ret) {
	    OsmoseCore core(argv[ret]);
	    core.run();
    } else {
            cout << "No ROM filename has been found on command line (.gg .GG .sms .SMS .zip .ZIP extensions)." << endl;
            usage();
    }

    return 0;    
}

/*--------------------------------------------------------------------*/
/* This method displays osmose usage.				      */
/*--------------------------------------------------------------------*/
void usage()
{
    cout << "Usage is osmose filename, ex: osmose asterix.zip -fs -tv -nosound" << endl << endl;
    cout << "Options: "<< endl;
    cout << "    -paddle           emulates one axis paddle (mapped on mouse)." << endl;
    cout << "    -acceleration x.x paddle acceleration (0.1 to 5 default "<< DEFAULT_ACCELERATION << ")" << endl;
    cout << "    -fs               run in fullscreen   (default: windowed)." << endl;
    cout << "    -nosound          do not play sounds. (default: sound on)."<< endl;
    cout << "    -dp               use dark palette for screen emulation (default: off)." << endl;
    cout << "    -inifile          xxx use xxx as configuration file. " <<endl;
    cout << "    -monochrom        emulates B&W tv (default: off)." << endl;
    cout << "    -scale2x          scale2x video filter implementation (default: off)." << endl;
    cout << "    -bilinear         bilinear video filter (default: off)." << endl;
    cout << "    -tv               emulates TV scanline video filter (default: off)." << endl;
    cout << "    -scale2xscanline  Scale2x + 75% scanline video filter (default: off)." << endl;
    cout << "    -nn2x             nearest neighbour video filter (default: off)." << endl;
    cout << "    -cm               use codemaster games mem. mapper (default: off). " << endl;
    cout << "    -km               use korean games mem. mapper (default: off). " << endl;
    cout << "    -irqhack          Enable irq hack (specific rom option. default off)." << endl;
    cout << "    -pal              emulates PAL/SECAM video timing (default: NTSC)."<< endl;
    cout << "    -jap              run as japanese sms (default: export)." << endl;
    cout << "    -exp              run as exported sms (default)." << endl;
    cout << "    -joy              use joystick as input device, instead of keyboard." << endl;
    cout << "    -joy1 x           define joystick button x as 1" << endl;
    cout << "    -joy2 x           define joystick button x as 2" << endl;
    cout << "    -joystart x       define joystick button x as Start (Game Gear only)" << endl;

}

/*--------------------------------------------------------------------*/
/* This method parses command line, and update options structure with */
/* right options.						      */
/*--------------------------------------------------------------------*/
int parseCommandLine(int a, char *rv[])
{
    int rom_filename_index = 0;

    for (int i=0; i<a; i++)
    {
        if (strcmp(rv[i],"-fs")==0)
        {
            emu_opt.fullscreen_flag = true;
        }

        if (strcmp(rv[i],"-pal")==0)
        {
            opt.ntsc = false;
        }

        if (strcmp(rv[i],"-fps")==0)
        {
            emu_opt.display_fps = true;
        }

        if (strcmp(rv[i],"-jap")==0)
        {
            // Act as Japanese machine.
            opt.WorldVersion = false;  
        }

        if (strcmp(rv[i],"-exp")==0)
        {
            // Act as Exported machine.
            opt.WorldVersion = true;
        }

        if (strcmp(rv[i],"-tv")==0)
        {
            emu_opt.videoFilterType = TV_FILTER;
        }

        if (strcmp(rv[i],"-scale2x")==0)
        {
            emu_opt.videoFilterType = SCALE2X_FILTER;
        }

        if (strcmp(rv[i],"-monochrom")==0)
        {
            emu_opt.videoFilterType = MONOCHROM_FILTER;
        }
        
        if (strcmp(rv[i],"-bilinear")==0)
        {
            emu_opt.videoFilterType = BILINEAR_FILTER;
        }

        if (strcmp(rv[i],"-scale2xscanline")==0)
        {
            emu_opt.videoFilterType = SCALE2XSCANLINE_FILTER;
        }

        if (strcmp(rv[i],"-nn2x")==0)
        {
            emu_opt.videoFilterType = NN2X_FILTER;
        }
        
        if (strcmp(rv[i],"-acceleration")==0)
        {
		    // If actual argument nbr < argument number eg is there any args ?
		    if (i < a-1)
		    {
	            opt.acceleration = atof( rv[i+1]);
				if (opt.acceleration == 0)
				{
			        cerr << "No valid acceleration parameter was given. Using default " << (float)DEFAULT_ACCELERATION << " value." << endl;
		            opt.acceleration = DEFAULT_ACCELERATION;
				}
				cout << "Paddle acceleration used : " << opt.acceleration << endl;
			}
			else
			{
			    cerr << "No acceleration parameter was given. Using default " << (float)DEFAULT_ACCELERATION << " value." << endl;
		        opt.acceleration = DEFAULT_ACCELERATION;
			}
        }

      
        if (strcmp(rv[i],"-nosound")==0)
        {
            emu_opt.sound = false;
        }

        if (strcmp(rv[i],"-cm")==0)
        {
            opt.mapperType = CodemasterMapper;
        }

        if (strcmp(rv[i],"-km")==0)
        {
            opt.mapperType = KoreanMapper;
        }

        if (strcmp(rv[i],"-irqhack")==0)
        {
            opt.irq_hack = true;
        }

        if (strcmp(rv[i],"-paddle")==0)
        {
            opt.inputType = PADDLE;
        }

        if (strcmp(rv[i],"-joy")==0)
        {
            opt.inputType = JOYSTICK;
        }

        if (strcmp(rv[i],"-joy1")==0)
        {
		   if (i < a-1)
	 	   {
	           emu_opt.joy1 = atoi( rv[i+1]);
				if (emu_opt.joy1 < 0)
				{
			        cerr << "No valid button for button 1 was given. Using default value." << endl;
	  	            emu_opt.joy1 = 0;
				}
				cout << "Button 1 mapped to joystick button " << emu_opt.joy1 << endl;
			}
			else
			{
			        cerr << "No valid button for button 1 was given. Using default value." << endl;
	  	            emu_opt.joy1 = 0;
			}
        }

        if (strcmp(rv[i],"-joy2")==0)
        {
		   if (i < a-1)
	 	   {
	           emu_opt.joy2 = atoi( rv[i+1]);
				if (emu_opt.joy2 < 0)
				{
			        cerr << "No valid button for button 2 was given. Using default value." << endl;
	  	            emu_opt.joy2 = 1;
				}
				cout << "Button 2 mapped to joystick button " << emu_opt.joy2 << endl;
			}
			else
			{
			        cerr << "No valid button for button 2 was given. Using default value." << endl;
	  	            emu_opt.joy2 = 1;
			}
        }

        if (strcmp(rv[i],"-joystart")==0)
        {
		   if (i < a-1)
	 	   {
	           emu_opt.joystart = atoi( rv[i+1]);
				if (emu_opt.joystart < 0)
				{
			        cerr << "No valid button for button Start was given. Using default value." << endl;
	  	            emu_opt.joystart = 2;
				}
				cout << "Button Start mapped to joystick button " << emu_opt.joystart << endl;
			}
			else
			{
			        cerr << "No valid button for button A was given. Using default value." << endl;
	  	            emu_opt.joystart = 2;
			}
        }


        if (strcmp(rv[i],"-dp")==0)
        {
            emu_opt.bright_palette = false;
        }

        // The ini file MUST be the next parameter after -inifile option.
        if (strcmp(rv[i],"-inifile")==0)
        {
		    // If actual argument nbr < argument number eg is there any args ?
		    if (i < a-1)
		    {
	            emu_opt.default_config = false;
		        strncpy(emu_opt.ini_file, rv[++i],512);
		    }
		    else
		    {
		        cerr << "No ini file was given, using default configuration." << endl;
	            emu_opt.default_config = true;
		    }
        }

        char extension [][5] = {".zip",".ZIP",".sms",".SMS",".gg",".GG"};
        for (int o = 0; o < 6; o++)
        {
                char *pattern = NULL;
                pattern = strstr(rv[i], extension[o]);
                if (pattern != NULL)
                {
                    rom_filename_index = i;
                }
        }

    }
    return(rom_filename_index);
}

/*--------------------------------------------------------------------*/
/* This method writes Osmose splash screen, version, author and build */
/* informations.						      */
/*--------------------------------------------------------------------*/
void splashText()
{
    cout << endl <<  __OSMOSE_VERSION__ <<" build on " << __DATE__<< " at " << __TIME__<< " Written by Bruno Vedder." << endl;
}

